File: Java/Memory.java
//File name: Memory.java
//Created by Joshua Brock on 9/24/2018

import java.util.ArrayList;

public class Memory
{

	private static ArrayList<Character> id = new ArrayList<Character>();
	private static ArrayList<Integer> mem = new ArrayList<Integer>();
	
	public static int fetch (char ch)
	{
		//Returns the data for a given character. Throws an exception if the given data isn't found.
		assert(getIndex(ch) != -1);
		return mem.get(getIndex(ch));
	}
	
	public static void store (char ch, int value)
	{
		//If character Id isn't present, add it and its value to our list, else override the old value.
		if(getIndex(ch) == -1)
		{
			id.add(ch);
			mem.add(value);
		}
		else
		{
			mem.set(getIndex(ch), value);
		}
	}
	
	private static int getIndex (char ch)
	{
		//Returns the index for a given character. Throws an exception if the given data isn't found.
		for(int i = 0; i < id.size(); i++)
		{
			if(id.get(i) == ch){
				return i;
			}
		}
		return -1;
	}
	
	public static void displayMemory()
	{
		//Returns every Id ever stored here and their value at the time of displaying. (in order)
		for (int i = 0; i < id.size(); i++)
			System.out.println (id.get(i) + ": " + mem.get(i));
	}
}

File: Python/Memory.py
'''File name: Memory.py
Created by Joshua Brock on 10/25/2018'''


idlst = []
mem = []


def fetch(ch):
    '''Precondition: ch is valid and is in our list
    Postcondition: returns the value associated with our ch'''
    assert (isinstance(ch, str) and getIndex(ch) != -1)
    return mem[getIndex(ch)]


def store(ch, value):
    '''Precondition: ch and value are valid
    Postcondition: creates new entries for ch and value, if ch already exists, replace its current value'''
    assert (isinstance(ch, str) and isinstance(value, int))
    if getIndex(ch) == -1:
        idlst.append(ch)
        mem.append(value)
    else:
        mem[getIndex(ch)] = value


def getIndex(ch):
    '''Precondition: ch is valid
    Postcondition: returns index of ch in our idlst. If not there returns -1'''
    assert (isinstance(ch, str))
    for x in range(len(idlst)):
        if idlst[x] == ch:
            return x
    return -1


def displayMemory():
    '''Postcondition: Returns all of the chars with their associated values'''
    for x in range(len(idlst)):
        print(str(idlst[x]) + ": " + str(mem[x]))
